README.md
Introduction
-----------------

XWRT (XML-WRT) is an efficient XML/HTML compressor (actually it works well with all textual files). 
It transforms XML to more compressible form and uses zlib (default), LZMA, PPMd, or lpaq6 as 
back-end compressor. This idea is based on well-known XML compressor: XMill.
Moreover XML-WRT creates a semi-dynamic dictionary and replaces frequently 
used words with shorter codes. There are additional techniques to improve
compression ratio:
- word alphabet can consist of start tags (like '<tag>'), urls, e-mails
- special model for numbers encoding
- input XML file is split into containers
- there are special containers for dates, time, pages and fractional numbers
- end tags ('</tag>') are replaced with a single char
- end tags + EOL symbols can also be replaced with a single char
- spaceless words model
- very effective methods for white-space preserving
- quotes modeling ('="' and '">' replaced with a single char) 


Comparision to other XML compressors
-------------------------

All files used for comparision can be downloaded from [Wratislavia XML Corpus]. Results are given in bpc (bits ber character). Tested with XWRT 3.1: 

|file       |gzip |XMill 0.9 zip|XWRT -l2 (zlib)|LZMA -a1|XWRT -l6 (LZMA)|PPMdJ -o8 -m64|XMill 0.9 PPMd|XMLPPM -l 9|SCMPPM -l 9|XWRT -l9 (PPM)|FastPAQ8 74 MB|XWRT -l11 (FastPAQ8)|
|-----------|-----|-------------|---------------|--------|---------------|--------------|--------------|-----------|-----------|--------------|--------------|--------------------|
|dblp       |1.463|        1.250|          0.865|   0.943|          0.747|         0.724|         0.940|      0.802|      0.693|         0.690|         0.659|               0.597|
|enwikibooks|2.339|        2.295|          1.742|   1.686|          1.504|         1.565|         1.838|      1.621|      1.621|         1.481|         1.357|               1.269|
|enwikinews |2.248|        2.198|          1.597|   1.462|          1.301|         1.291|         1.746|      1.379|      1.398|         1.202|         1.172|               1.090|
|lineitem   |0.721|        0.380|          0.276|   0.421|          0.243|         0.359|         0.270|      0.261|      0.242|         0.243|         0.236|               0.226|
|Shakespeare|2.182|        2.044|          1.481|   1.646|          1.349|         1.245|         1.584|      1.295|      1.293|         1.204|         1.220|               1.185|
|SwissProt  |0.985|        0.619|          0.475|   0.478|          0.388|         0.490|         0.477|      0.416|      0.417|         0.363|         0.395|               0.313|
|uwm        |0.553|        0.382|          0.315|   0.368|          0.278|         0.426|         0.310|      0.259|      0.274|         0.240|         0.254|               0.228|
|**average**|**1.499**|**1.310**|**0.964**|**1.001**|**0.830**|**0.871**|**1.024**|**0.862**|**0.848**|**0.775**|**0.756**|**0.701**|

[Wratislavia XML Corpus]: http://pskibinski.pl/research/Wratislavia/



Usage
-----------------

```
Usage: XWRT.exe [options] <file1> [file2] [file3] ...
 where <file> is a XML file or a XWRT compressed file (it's auto-detected)
             you can also use wildcards (e.g., "*.xml")

GENERAL OPTIONS (which also set default additional options):
  -l0 = no compression (memory usage up to 16 MB)
  -l1 = zlib fast (memory usage 16+1 MB)
  -l2 = zlib normal (default, memory usage 16+1 MB)
  -l3 = zlib best (memory usage 16+1 MB)
  -l4 = LZMA dict size 64 KB (memory usage 16+9 MB for compression and 16+3 MB for decompression)
  -l5 = LZMA dict size 1 MB (memory usage 16+18 MB for compression and 16+3 MB for decompression)
  -l6 = LZMA dict size 8 MB (memory usage 16+84 MB for compression and 16+10 MB for decompression)
  -l7 = PPMd model size 16 MB (memory usage 16+20 MB)
  -l8 = PPMd model size 32 MB (memory usage 16+36 MB)
  -l9 = PPMd model size 64 MB (memory usage 16+70 MB)
  -l10 = lpaq6 model size 120 MB (memory usage 16+104 MB)
  -l11 = lpaq6 model size 214 MB (memory usage 16+198 MB)
  -l12 = lpaq6 model size 406 MB (memory usage 16+390 MB)
  -l13 = lpaq6 model size 790 MB (memory usage 16+774 MB)
  -l14 = lpaq6 model size 1560 MB (memory usage 16+1542 MB)
  -0 = preprocessed and uncompressed output optimized for further LZ77 compression
  -1 = preprocessed and uncompressed output optimized for further LZMA compression
  -2 = preprocessed and uncompressed output optimized for further PPM compression
  -3 = preprocessed and uncompressed output optimized for further PAQ compression

ADDITIONAL OPTIONS:
  -bX = Set maximum buffer size while creating dynamic dictionary to X MB
  -c = Turn off containers (without number and word containers)
  +d = Turn on usage of the static dictionary (requires wrt-eng.dic,
       which is available at http://pskibinski.pl/research)
  -eX = Set maximum dictionary size to X words
  -fX = Set minimal word frequency to X
  -i = Delete input files
  -mX = Set maximum memory buffer size to X MB (default=8)
  -n = Turn off number containers
  -o = Force overwrite of output files
  -pX = Preprocess only (file_size/X) bytes in a first pass
  -s = Turn off spaces modeling option
  -t = Turn off "try shorter word" option
  -w = Turn off word containers
```


Compilation
-------------------------
For Linux/Unix:
```
make BUILD_SYSTEM=linux
```

For Windows (MinGW)
```
make
```


Used libraries
---------------------
```
zlib  (C) 1995-2005 Jean-loup Gailly and Mark Adler 
LZMA  (C) 1999-2006 Igor Pavlov
PPMd (C) 1997-2006 Dmitry Shkarin
lpaq6  (C) 2007 Matt Mahoney and Alexander Ratushnyak
```


HISTORY.md
XWRT 3.4 (5.11.2007)
-PPMVC replaced with PPMd var J1
-added support for LZMA and PPMd back-end compression on Linux

XWRT 3.2 (25.10.2007)
-FastPAQ8 replaced with lpaq6 (compression level 10-14)

XWRT 3.1 (05.06.2007)
-improved support for XML files encoded in UTF-8
-dictionary is compressed using front compression
-added little-endian/big-endian Unicode (UCS-2) support
-non-textual files are compressed/stored without using a filter
-64-bit compiler support

XML-WRT 3.0 (14.09.2006)
-internal PPMVC and FastPAQ8 compression

XML-WRT 2.0 (14.06.2006)
-internal zlib and LZMA compression
-input XML file is split into containers depend on start-tags and end-tags and content under the same tag is sent to the same container
-container for dates in format 1980-02-31 and 01-MAR-1920
-container for times in format 11:30pm
-container for numbers from 1900 to 2155 (years)
-container for pages in format "x-y", where y-x<256, eg. "120-148", "1480-1600"
-container for numbers in format "x-y", eg. "1234-0", "87-623"
-container for two digits after period, eg. "102.00", "12.01"
-container for numbers from 0.0 to 24.9 (one digit after period), eg. "12.0", "9.9"
-urls (statring from "http:"), e-mails (x@y.z), "&uuml;" added to dynamic dictionary

XML-WRT 1.0 (27.03.2006)
-first public release


src/XWRT.cpp
#include "XWRT.h"
#include <time.h>
#include <vector>
#include "Encoder.h"
#include "Decoder.h"

#ifdef WINDOWS
	#include <windows.h> // createFileList
	#include <conio.h> // getch()
#else
	#define getch getchar
#endif

unsigned int bytesRead=0,bytesWritten=0; //supports up to 4GB
int g_fileLenMB;
bool YesToAll=false;

void printStatus(int add_read,int add_written,int encoding) 
{  // print progress
	bytesRead+=add_read;
	bytesWritten+=add_written;
	if (encoding)
		PRINT_STATUS(("%10u (%3d%%) -> %10u \b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b", bytesRead,(g_fileLenMB>0)?(bytesRead/1024/1024*100)/g_fileLenMB:0,bytesWritten))
	else
		PRINT_STATUS(("%10u -> %10u (%3d%%) \b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b", bytesRead,bytesWritten,(g_fileLenMB>0)?(bytesWritten/1024/1024*100)/g_fileLenMB:0));
}


bool tryOpen(const char* FName, bool& YesToAll)
{
    FILE* fp=fopen(FName,"rb");
    if (!fp)                              
		return true;
    fclose(fp);

    if (YesToAll)
	{
		remove(FName);
		return true;
	}

    printf("%s already exists, overwrite?: <Y>es, <N>o, <A>ll, <Q>uit?",FName);
    for ( ; ; )
        switch ( toupper(getch()) ) {
            case 'A':
				YesToAll=true;
            case '\r': 
			case 'Y':
				remove(FName);
				return true;
            case 0x1B: 
			case 'Q':
				printf("quit\n"); exit(-1);
            case 'N':                       
				return false;
        }
}

void start_encode(FILE* file,char* filename,bool WRT_verbose,int argc, char* argv[])
{
	FILE* fileout;
	char fileoutbuf[32*1024];
	clock_t start_file;  // in ticks

	bytesRead=bytesWritten=0;

	XWRT_Encoder xml_wrt;
	xml_wrt.YesToAll=YesToAll;
	xml_wrt.WRT_verbose=WRT_verbose;
	xml_wrt.defaultSettings(argc,argv);
	
	start_file=clock();
	
	std::string outputFilename=filename;
	outputFilename+=ADD_FILENAME_EXT;
	
	if (!tryOpen(outputFilename.c_str(),xml_wrt.YesToAll))
		return;
	
#ifdef USE_LZMA_LIBRARY
	int preprocFlag=xml_wrt.preprocFlag;
	if (IF_OPTION(OPTION_LZMA))
	{
		if (!LZMAlib_PrepareOutputFile((char*)outputFilename.c_str(),xml_wrt.outStream))
		{
			printf("Can't open file %s\n",outputFilename.c_str());
			return;
		}
		fileout=(FILE*)1;
	}
	else
#endif
	{
		fileout=fopen(outputFilename.c_str(),"wb");
		if (fileout==NULL)
		{
			printf("Can't open file %s\n",outputFilename.c_str());
			fclose(file);
			return;
		}
		setvbuf(fileout, fileoutbuf, _IOFBF, sizeof(fileoutbuf));
	}
			
	std::string compName;
	xml_wrt.getAlgName(compName);
	printf("- encoding %s to %s (%s)\n",filename,outputFilename.c_str(),compName.c_str());
					
	int flen=xml_wrt.flen(file);
	XWRT_file=file;
	XWRT_fileout=fileout;
	fseek(XWRT_file, 0, SEEK_SET );
	xml_wrt.WRT_start_encoding(flen,false);
				
	if (WRT_verbose)
	{
		float tim=float(clock()-start_file)/CLOCKS_PER_SEC;
#ifdef USE_LZMA_LIBRARY
		if (IF_OPTION(OPTION_LZMA))
			printf(" + encoding finished (%d->%d bytes, %.03f bpc) in %.02fs (%.0f kb/s)\n", (int)ftell(file), LZMAlib_GetOutputFilePos(xml_wrt.outStream),LZMAlib_GetOutputFilePos(xml_wrt.outStream)*8.0/ftell(file),tim,ftell(file)/1024/tim);
		else
#endif
			printf(" + encoding finished (%d->%d bytes, %.03f bpc) in %.02fs (%.0f kb/s)\n", (int)ftell(file), (int)ftell(fileout), ftell(fileout)*8.0/ftell(file), tim, ftell(file)/1024.0/tim);
	}
				
	fclose(file);


#ifdef USE_LZMA_LIBRARY
	if (IF_OPTION(OPTION_LZMA))
		LZMAlib_CloseOutputFile(xml_wrt.outStream);
	else
#endif
		fclose(fileout);

		
	if (xml_wrt.deleteInputFiles)
		remove(filename);

	if (!YesToAll)
		YesToAll=xml_wrt.YesToAll;
}

void start_decode(FILE* file,char* filename,bool WRT_verbose,int argc, char* argv[])
{
	FILE* fileout;
	char fileoutbuf[32*1024];
	clock_t start_file;  // in ticks

	bytesRead=bytesWritten=0;

	XWRT_Decoder xml_wrt;
	xml_wrt.YesToAll=YesToAll;
	xml_wrt.WRT_verbose=WRT_verbose;
	xml_wrt.defaultSettings(argc,argv);

	start_file=clock();
	
	if (getc(file)!=XWRT_HEADER[3])
	{
		printf("Bad XWRT version!\n");
		return;
	}

    int c;
	if ((c=getc(file))!=XWRT_VERSION-150)
	{
        c += 150;
		printf("Bad XWRT version, you need XWRT %d.%d!\n", c/100, (c/10)%10);
		return;
	}

	c=getc(file);
	int preprocFlag=0;
	if ((c&8)!=0)
		TURN_ON(OPTION_LZMA);

	std::string outputFilename=filename;
	size_t pos=outputFilename.rfind(CUT_FILENAME_CHAR);
	if (pos!=std::string::npos && pos>0)
		outputFilename.resize(pos);
	else
		outputFilename+=".xml";
	
	if (!tryOpen(outputFilename.c_str(),xml_wrt.YesToAll))
		return;
	
	fileout=fopen(outputFilename.c_str(),"wb");
	if (fileout==NULL)
	{
		printf("Can't open file %s\n",outputFilename.c_str());
		fclose(file);
		return;
	}
	setvbuf(fileout, fileoutbuf, _IOFBF, sizeof(fileoutbuf));
	
	
#ifdef USE_LZMA_LIBRARY
	if (IF_OPTION(OPTION_LZMA))
	{
		fclose(file);
		if (!LZMAlib_PrepareInputFile(filename,xml_wrt.inStream))
		{
			printf("Can't open file %s\n",outputFilename.c_str());
			return;
		}	

		for (int i=0;i<6; i++)
			c=LZMAlib_getc(xml_wrt.inStream);  // XWRT header + ver + 1 byte
	}
#endif

	XWRT_file=file;
	XWRT_fileout=fileout;
	xml_wrt.WRT_start_decoding(c,filename,(char*)outputFilename.c_str());
	
	if (WRT_verbose)
	{
		float tim=float(clock()-start_file)/CLOCKS_PER_SEC;

#ifdef USE_LZMA_LIBRARY
		if (IF_OPTION(OPTION_LZMA))
			printf(" + decoding finished (%d->%d bytes, %.03f bpc) in %.02fs (%.0f kb/s)\n",LZMAlib_GetInputFilePos(xml_wrt.inStream),(int)ftell(fileout),LZMAlib_GetInputFilePos(xml_wrt.inStream)*8.0/ftell(fileout),tim,ftell(fileout)/1024/tim);
		else
#endif
			printf(" + decoding finished (%d->%d bytes, %.03f bpc) in %.02fs (%.0f kb/s)\n", (int)ftell(file), (int)ftell(fileout), ftell(file)*8.0/ftell(fileout), tim, ftell(fileout)/1024/tim);
	}

	fclose(fileout);

#ifdef USE_LZMA_LIBRARY
	if (IF_OPTION(OPTION_LZMA))
		LZMAlib_CloseInputFile(xml_wrt.inStream);
	else
#endif
		fclose(file);

	
		
	if (xml_wrt.deleteInputFiles)
		remove(filename);

	if (!YesToAll)
		YesToAll=xml_wrt.YesToAll;
}



void processFile(char* filename,bool WRT_verbose,int argc, char* argv[])
{
	FILE* file;
	char filebuf[32*1024];

	file=fopen((const char*)filename,"rb");
	if (file==NULL)
	{
		printf("Can't open file %s\n",filename);
		return;
	}
	
	setvbuf(file, filebuf, _IOFBF, sizeof(filebuf));
	
	
	if (getc(file)==XWRT_HEADER[0] && getc(file)==XWRT_HEADER[1] && 
		getc(file)==XWRT_HEADER[2])
		start_decode(file,filename,WRT_verbose,argc,argv);
	else
		start_encode(file,filename,WRT_verbose,argc,argv);
}

#ifdef WINDOWS

std::vector<std::string> list;

bool createFileList(char* pattern)
{
	std::string str;
	std::string patt;
    WIN32_FIND_DATA c_file;
    HANDLE hFile=FindFirstFile( pattern, &c_file );
	patt=pattern;

	if (hFile != INVALID_HANDLE_VALUE) 
	do
	{
		str=c_file.cFileName;

		if (!(c_file.dwFileAttributes & FILE_ATTRIBUTE_DIRECTORY))
			list.push_back(str);
	}
	while (FindNextFile( hFile, &c_file ));

	FindClose( hFile );

	return true;
}

#endif

int getOptionsCount(int argc, char* argv[])
{
	int optCount=0;

	while (argc>1 && (argv[1][0]=='-' || argv[1][0]=='+')) 
	{
		argc--;
		optCount++;
		argv++;
	}

	return optCount;
}

void usage()
{
	printf("Usage: XWRT.exe [options] <file1> [file2] [file3] ...\n");
	printf(" where <file> is a XML file or a XWRT compressed file (it's auto-detected)\n");
	printf(" 	     you can also use wildcards (e.g., \"*.xml\")\n");
	printf(" GENERAL OPTIONS (which also set default additional options):\n");
	printf("  -i = Delete input files\n");
	printf("  -lX = Set compression level (0=store");
#ifdef USE_ZLIB_LIBRARY
	printf("; 1,2,3=zlib [fast,normal(default),best]");
#endif
#if defined(USE_LZMA_LIBRARY) || defined(USE_PPMD_LIBRARY) || defined(USE_PAQ_LIBRARY)
	printf(";\n     ");
#endif
#ifdef USE_LZMA_LIBRARY
	printf("4,5,6=LZMA [64 KB, 1 MB, 8 MB]; ");
#endif
#ifdef USE_PPMD_LIBRARY
	printf("7,8,9=PPMd [16 MB, 32 MB, 64 MB];\n     ");
#endif
#ifdef USE_PAQ_LIBRARY
	printf("10,11,12,13,14=lpaq6 [104 MB, 198 MB, 390 MB, 774 MB, 1542 MB]");
#endif
	printf(")\n");
	printf("  -o = Force overwrite of output files\n");
	printf("  -0, -1, -2, -3 = -l0 (store) optimized for further LZ77, LZMA, PPM, PAQ (-3)\n");
	printf(" ADDITIONAL OPTIONS:\n");
	printf("  -bX = Set maximum buffer size while creating dynamic dictionary to X MB\n");
	printf("  -c = Turn off containers (without number and word containers)\n");
	printf("  +d = Turn on usage of the static dictionary (requires wrt-eng.dic,\n");
	printf("       which is available at http://pskibinski.pl/research)\n");
	printf("  -eX = Set maximum dictionary size to X words\n");
	printf("  -fX = Set minimal word frequency to X\n");
	printf("  -mX = Set maximum memory buffer size to X MB (default=8)\n");
	printf("  -n = Turn off number containers\n");
	printf("  -pX = Preprocess only (file_size/X) bytes in a first pass\n");
	printf("  -s = Turn off spaces modeling option\n");
	printf("  -t = Turn off \"try shorter word\" option\n");
	printf("  -w = Turn off word containers\n");
}


int main(int argc, char* argv[])
{
	int optCount;
	bool WRT_verbose=true;

	if (WRT_verbose)
		printf(PRG_NAME "\n");

	clock_t start_time;  // in ticks
    start_time=clock();

	optCount=getOptionsCount(argc,argv);


	if (argc-optCount<2)
	{
		usage();
		return 0;
	};

	int curr=optCount+1;

	while (curr<argc)
	{
#ifdef WINDOWS
		createFileList(argv[curr]);
#else
		processFile(argv[curr], WRT_verbose, argc, argv);
#endif
		curr++;
	}	

#ifdef WINDOWS
	std::vector<std::string>::iterator it;
	if (list.size()<=0)
		printf("Can't open input file(s)\n");
	else
	for (it=list.begin(); it!=list.end(); it++)
		processFile((char*)it->c_str(), WRT_verbose, argc, argv);
#endif

	if (argc>optCount+2)
	    printf("+ Total time %1.2f sec\n",double(clock()-start_time)/CLOCKS_PER_SEC);

	return 0;
};

